Mission API

WatchTheFort opened the pull request:
### Work done
@Psyborg8's work done on the mission API:
- Initial OOP architecture to support using classes for various object types
-  Implement additional triggers and actions
- Refactor trigger and action parameter checking code
- Lots of code cleanup

#### Test steps
- [ ] Enable the main mission API gadget `api_missions`
- [ ] Start up a skirmish against an InactiveAI
- [ ] Check that the triggers and actions in `singleplayer/test_mission.lua` are hit

Psyborg8 committed:
Optimized gui_game_type_info.lua
- Removed all allocations and string concatenation made in widget:DrawScreen, the messages never change after game start so this was unnecessary
- Moved scaling calculations  from widget:DrawScreen to widget:ViewResize. This is the only time the size will change

Reduced runtime allocation to 0, from ~1500 kB/s, and reduced usage by ~1%

Psyborg8 committed:
Slight refactor and fixed blink
- Fixed styling
- Blink was repeatedly reformatting string each frame unnecessarily

Psyborg8 committed:
Update GameTypeInfo strings on LanguageChanged

Psyborg8 committed:
Cleaned up Initialize

Psyborg8 committed:
Disabled blink timer when blinked message isn't used

Psyborg8 committed:
- Beautification (headers and separators)
- spawnUnits and despawnUnits now handle quanitity

Psyborg8 committed:
Merge branch 'beyond-all-reason:master' into Mission-API

Psyborg8 committed:
- Implemented IssueOrders action
- Added validation checking for unit, unitDef, vec3 and direction parameter types
- Added actions_defs.lua for storing parameter enums

Psyborg8 committed:
Merge branch 'Mission-API' of https://github.com/Psyborg8/Beyond-All-Reason into Mission-API

Psyborg8 committed:
Fixed a typo

Psyborg8 committed:
- Implemented TransferUnits, SpawnExplosion
- TrackedUnits is now always an array to simplify implementations, and allow single units to be added to a tracked name

Psyborg8 committed:
Merge branch 'beyond-all-reason:master' into Mission-API

Psyborg8 committed:
Cleaned up trigger code
Implemented UnitExists, UnitNotExists, UnitKilled, UnitCaptured, UnitEnteredLocation, ConstructionStarted and Construction Finished triggers
Implemented Custom action

Psyborg8 committed:
Merge branch 'Mission-API' of https://github.com/Psyborg8/Beyond-All-Reason into Mission-API

Psyborg8 committed:
Implemented OO solution to trigger and action parameters

Psyborg8 committed:
Merge branch 'beyond-all-reason:master' into Mission-API

Psyborg8 committed:
types.lua documentation added

WatchTheFort committed:
Merge pull request #2802 from Psyborg8/Mission-API

Mission API dev work from @Psyborg8's fork.

RebelNode commented:
This branch is giving me `[LuaVFS::Include(synced=true)][pcall] file=luarules/mission_api/types.lua error=2 ([string "luarules/mission_api/types.lua"]:7: attempt to index field 'tracker' (a nil value)) ptop=1 cenv=false vfsmode=Mmb) ptop=3 cenv=false vfsmode=Mmb` resulting in `LuaRules is not loaded` when enabling mission api.

WatchTheFort commented:
This code has not actually been tested yet, there is still a lot of work to do.

RebelNode commented:
Yeah I can see that. I'll try to test it. I got over this error with some changes but didn't get the code to actually run the triggers/actions yet.

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Linting: Convert spaces to tabs

WatchTheFort committed:
Linting: Standardize capitalization

WatchTheFort committed:
Rename Collider -> ProximityMonitor, rename some properties for clarity

Ezahiel committed:
Minor bugs

Ezahiel committed:
Merge branch 'beyond-all-reason:mission-api/dev' into mission-api/dev

Ezahiel committed:
disabled tracker, msg working (#3229)

Minor bugs

WatchTheFort committed:
Remove references to modoptions to not bias dev direction

Ezahiel committed:
initial work on OO Actions / Validation

Ezahiel committed:
validation update & direction rework

WatchTheFort committed:
Instantiate when using custom types

Ezahiel committed:
Mission api/dev (#3341)

* Minor bugs

* initial work on OO Actions / Validation

* validation update & direction rework

WatchTheFort committed:
Instantiate when using custom types

WatchTheFort committed:
Fix merging mistakes

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Fix Vec2 and Direction constructors, calling SpawnUnits action

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Mission API: Add victory and defeat actions (#3662)

* Mission API: add initial victory and defeat actions, they grant victory to all ally teams having players

S3KCentrifugal commented:
Long term versioning could become an issue if the API outgrows the scenarios. Perhaps add a min version to scenario? Or find a way to add scenario tests?

What additional steps are required to get this supporting multiplayer?

sorenmarkert committed:
Mission API: Victory and Defeat actions take winners/losers as args (#4188)

Modified the Victory and Defeat actions so instead of trying to guess who's the winning ally team(s), let the
scenario author provide their IDs to the actions.

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Remove object-oriented code

WatchTheFort committed:
Change parameter type from unit to string

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Replace parameter custom types with primitive types

WatchTheFort committed:
Fix position paramater type string -> table

WatchTheFort committed:
Cleanup comments

sorenmarkert committed:
Mission API: Fix actions SpawnUnits, DespawnUnits, and IssueOrders (#5785)

- Fixed the mission API actions SpawnUnits, DespawnUnits, and IssueOrders.
- Made SpawnUnits spawn units in different locations instead of all in the same spot.
- Added more testing.

sorenmarkert committed:
Mission API: Fix TransferUnits (#6589)

- TransferUnits action now acts on units by name.
- Added tests.

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Mission API: Add actions NameUnits and UnnameUnits (#6599)

Added new actions 'NameUnits' and 'UnnameUnits'

sorenmarkert committed:
Mission API: Custom validation functions (#6696)

- Added custom parameter validation.
- This moves the validation of Lua types from the two loaders to functions in a shared validation.lua.
- It also adds further validation of "sub-types" of the Lua types, such as unitDefName, position, teamID, etc.

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Mission API: Improve TransferUnits and SpawnUnits actions (#6966)

- Removed the `given` parameter from TransferUnits, calculating the value instead.
- Added a `spacing` parameter to SpawnUnits

sorenmarkert committed:
Add unit triggers

sorenmarkert committed:
Mission API: Add Marker Actions (#6697)

Added actions AddMarker, EraseMarker, and DrawLines.

sorenmarkert committed:
Activate trigger for each unit that entered or left an area

sorenmarkert committed:
Add stages

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Add objectives

sorenmarkert committed:
Add triggers and actions for handling features

sorenmarkert committed:
Add loadouts and SpawnLoadout action

sorenmarkert committed:
Mission API: Add PlaySound action (#6964)

* Add the PlaySound action. It plays .wav files with an option for queued play for voice lines, which also pauses the normal voice notifications.
* Refactor parameter to an explicit step in the trigger/action loading process

sorenmarkert committed:
Mission API: Update SpawnExplosion action  (#6979)

SpawnExplosion action now needs only a weaponDefName and a position to work.

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/add-unit-triggers

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions.lua
#	luarules/mission_api/validation.lua

sorenmarkert committed:
Update validation.lua after merging refactoring

sorenmarkert committed:
Add resource actions and triggers

sorenmarkert committed:
Revert "Add resource actions and triggers"

This reverts commit 10bc82e4271feb33069ceaf01470f16ff102a3c6.

sorenmarkert committed:
Add UI widget

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/stages-and-objectives

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/validation.lua
#	singleplayer/mission-api-tests/validation_test.lua

sorenmarkert committed:
Fix untracking bug

sorenmarkert committed:
Fix untracking bug

sorenmarkert committed:
Validate stage title and objective text don't contain |

sorenmarkert committed:
Refactor tracking to use tables of sets instead of tables of arrays

sorenmarkert committed:
Merge branch 'mission-api/add-unit-triggers' into mission-api/add-feature-triggers-and-actions

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions.lua
#	luarules/mission_api/parameter_types.lua
#	luarules/mission_api/tracking.lua
#	luarules/mission_api/validation.lua
#	singleplayer/mission-api-tests/validation_test.lua

sorenmarkert committed:
Update after merge

sorenmarkert committed:
Fix typo

sorenmarkert committed:
Review comment by sprunk: GiveOrderArrayToUnitMap

sorenmarkert committed:
Fix bad merge

sorenmarkert committed:
Add orders to loadouts and fix empty

sorenmarkert committed:
Finish resurrection in CreateFeature action and UnitResurrected trigger

sorenmarkert committed:
Merge branch 'mission-api/add-feature-triggers-and-actions' into mission-api/loadout

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions.lua
#	luarules/mission_api/actions_dispatcher.lua
#	luarules/mission_api/actions_schema.lua
#	luarules/mission_api/validation.lua
#	singleplayer/mission-api-tests/validation_test.lua

sorenmarkert committed:
fix merge

sorenmarkert committed:
Replace resurrectAs with calculated unitDefName reused from CreateFeature action

sorenmarkert committed:
Mission API: Add unit triggers (#6972)

- Extracted handling of tracked units to separate file
- Refactored trigger implementation
- Added all unit related triggers

sorenmarkert committed:
Merge branch 'refs/heads/mission-api/dev' into mission-api/stages-and-objectives

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/gadgets/api_missions_triggers.lua

sorenmarkert committed:
Merge branch 'refs/heads/mission-api/dev' into mission-api/add-feature-triggers-and-actions

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/gadgets/api_missions_triggers.lua
#	luarules/mission_api/actions.lua
#	luarules/mission_api/tracking.lua
#	luarules/mission_api/triggers_schema.lua
#	singleplayer/mission-api-tests/unit_triggers_test.lua
#	singleplayer/mission-api-tests/validation_test.lua

sorenmarkert committed:
Fix bad merge, test, and validation

sorenmarkert committed:
Update after merge

sorenmarkert committed:
Merge branch 'refs/heads/mission-api/add-feature-triggers-and-actions' into mission-api/loadout

# Conflicts:
#	luarules/mission_api/actions.lua

sorenmarkert committed:
Make stages and objectives optional, and init widget on :GamePreload

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/add-feature-triggers-and-actions

# Conflicts:
#	common/tablefunctions.lua

sorenmarkert committed:
Review comment by efrec on (#7018): simpler default value

sorenmarkert committed:
Review comment by efrec: comments, reclaim check, remove unused parm

sorenmarkert committed:
Use new table.ensureTable()

sorenmarkert committed:
Mission API: Give orders to named units (#6965)

The IssueOrders action now accepts orders containing a unitName as parameter. Such orders are then split into a series of orders for the names units.

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/add-feature-triggers-and-actions

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions.lua

sorenmarkert committed:
Review comment by efrec: reclaim check

sorenmarkert committed:
Review comment by WTF: param ordering

sorenmarkert committed:
Review comment by WTF on (#6965): IssueOrders works on feature names

sorenmarkert committed:
Merge branch 'mission-api/add-feature-triggers-and-actions' into mission-api/loadout

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/validation.lua

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/stages-and-objectives

# Conflicts:
#	luarules/gadgets/api_missions.lua

sorenmarkert committed:
Review comments by WTF: refactoring, renaming, removals

sorenmarkert committed:
Review comments by WTF: Add feature flag guard

sorenmarkert committed:
Review comments by WTF: utility functions, and renaming

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/add-feature-triggers-and-actions

sorenmarkert committed:
Mission API: Add feature triggers and actions (#7018)

- Added actions CreateFeature and DestroyFeature.
- Added triggers FeatureCreated, FeatureReclaimed, and FeatureDestroyed.

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/add-feature-triggers-and-actions

sorenmarkert committed:
Merge branch 'mission-api/add-feature-triggers-and-actions' into mission-api/loadout

# Conflicts:
#	luarules/mission_api/actions.lua

sorenmarkert committed:
clean up after merge

sorenmarkert committed:
Add string.facingToHeading and use it

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/loadout

sorenmarkert committed:
Update after merging dev

sorenmarkert committed:
Update after merging dev

sorenmarkert committed:
Review comments by WTF: various removals, use type index for action functions

sorenmarkert committed:
Review comments by WTF: SpawnLoadout moved to into SpawnUnits + CreateFeatures, fix duplication.

sorenmarkert committed:
Invalid references logged as warnings, and remove gadgets on validation errors.

sorenmarkert committed:
Mission API: Add statistics triggers (#7252)

Added triggers TotalUnitsLost, TotalUnitsBuilt, TotalUnitsKilled, TotalUnitsCaptured, and UnitsOwned.

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/loadout

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions.lua
#	luarules/mission_api/actions_schema.lua
#	luarules/mission_api/validation.lua

sorenmarkert committed:
Review comments by WTF: Remove unnecessary UnitDefNames, update comment

sorenmarkert committed:
Review comments by WTF: gadget validation check

sorenmarkert committed:
Add `/luarules dumploadout`

sorenmarkert committed:
Mission API: Add resource actions and triggers (#7251)

Added actions and triggers for team resources.

sorenmarkert committed:
Merge branch 'mission-api/dev' into mission-api/loadout

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions.lua
#	luarules/mission_api/actions_dispatcher.lua
#	luarules/mission_api/parameter_types.lua
#	luarules/mission_api/validation.lua

sorenmarkert committed:
Update resource_test.lua to use unitLoadout and featureLoadout

WatchTheFort committed:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Mission API: Add loadouts and SpawnLoadout action (#7250)

- Added unit and feature loadouts to mission scripts.
- Added SpawnLoadout action.

sorenmarkert committed:
Merge branch 'mission-api/loadout' into mission-api/stages-and-objectives

# Conflicts:
#	luarules/gadgets/api_missions.lua
#	luarules/mission_api/actions_dispatcher.lua
#	luarules/mission_api/validation.lua
#	singleplayer/mission-api-tests/validation_test.lua

sorenmarkert committed:
Update after merging loadouts

WatchTheFort committed:
Merge branch 'mission-api/dev' into mission-api/stages-and-objectives

WatchTheFort committed:
Missed a comma

sorenmarkert committed:
Review comments by WTF: Parameter types, and removals

sorenmarkert committed:
Remove widget and broadcast, fix trigger checks

sorenmarkert committed:
Remove quantity from UnitExists trigger, and make 'unitDefName' optional

sorenmarkert committed:
Refactor objectives: inline trigger and schema

sorenmarkert committed:
Review comments by WTF: Rename parm, change parm types, fix spec run

sorenmarkert committed:
Review comments by WTF: Revert optional trigger parms

sorenmarkert committed:
Review comments by WTF: Move table.copy

sorenmarkert committed:
Review comments by WTF: Replace objective update callback with a lookup table

sorenmarkert committed:
Update validation spec: textKey, loadouts, and references

sorenmarkert committed:
Review comments by WTF: Replace global with callback for unit name tracking

sorenmarkert committed:
Revert "Review comments by WTF: Replace global with callback for unit name tracking"

This reverts commit 98ace8890d0db101b423e3b349eb6ccbd8d21bc6.

sorenmarkert committed:
Review comments by WTF: Renames and comments

sorenmarkert committed:
Review comments by WTF: Move stages from objectives to top level of mission and update loading, validation, and tests

sorenmarkert committed:
Review comments by WTF: Refactor tryAdvanceStage() remove global ObjectiveToStages

sorenmarkert committed:
Review comments by WTF: Do all validation after pre-processing

sorenmarkert committed:
Review comments by WTF: Replace derivedStages with direct reference to GG['MissionAPI'].Stages

sorenmarkert committed:
Remove unused resetStagesCache function

sorenmarkert committed:
Move getTypesWithParameterType to util, various cleanup

sorenmarkert committed:
Review comments by WTF: tryAdvanceStage considers only objectives with same nextStage

WatchTheFort committed:
Mission API: Add objectives and stages (#7004)

This adds:

- ChangeStage action
- Objectives, a top level mission element with an inline trigger, and a set of stages
- UpdateObjective action

sorenmarkert committed:
Mission API: Refactor Actions (#8226)

Refactor actions into a separate file for each, containing

- Action type
- Schema
- Implementation function

These action files are loaded dynamically.

Also, this makes shared modules load only once so as to not take up memory multiple times. See [this chat](https://discord.com/channels/549281623154229250/724924957074915358/1524069585702097059) for reference.

Damgam committed:
Added support for Action Files to contain group of actions (#8433)

Required by
https://github.com/beyond-all-reason/Beyond-All-Reason/pull/8430

Allows to contain multiple actions per action file, to group similar
actions together.
All actions, even if there's only one, are consistently nested as table
within table, to reduce loader code complexity.

Damgam committed:
Hotfix for Multiple Actions Per File. (#8436)

Damgam committed:
Created set of Kill Unit actions. (#8430)

- DestroyUnits - Kills the units, with death animation and wreck. 
- SelfDestructUnits - Kills the units, with self-destruction explosion.
- ReclaimUnits - Despawns units with no explosion, gives metal equal to units metal cost to the killerTeam, just like if it was reclaimed.
- DespawnUnits - Despawns units with no explosion, doesn't give resources back.

Updated all occurences of old DespawnUnits action in test missions
Updated unit test which was failing due to removed parameter.

Damgam committed:
Merge branch 'master' into mission-api/dev

Damgam committed:
Mission API - Play Music (#8508)

- PlayMusic action
- Accepts filepath to specific music track
- No validation yet like with Sound Effects.
- No special functionality like queueing or smooth fade for now - TBD - this requires bigger refactor of music player side.
- Updated sound_test mission to play music, tested, works perfectly.

sorenmarkert committed:
Mission API: Refactor Triggers (#8450)

1. Triggers were split up into 1 file each.
2. Statistics handling was extracted to its own module.

Damgam committed:
Mission API - Added Resource Actions (#8506)

- Existing addResources action now supports removing resources as well
- Added new addResourcesPerSecond action
- Added Mission API Actions Helper gadget to handle resources per second for now until there's better solution.
- Adjusted the resource_test mission to test all the new features.

Ruwetuin committed:
Mission API - add action: unit selectable (#8547)

efrec committed:
Mission API: add EnumSets (#8580)

### Work done

Constructs EnumSet entries from base specs for shared validation and type handling. `EnumSets` otherwise still works like `Enums`, and each enumSet is included in `Enums`.

The API should keep identical expectations around set-like behavior (nil is permissive, no empty tables allowed, specifying any value in the set excludes unspecified values).

Damgam committed:
Mission API Action files SubDirs (#8498)

- Prepared Actions Folder for subdirectories
- SubDirs are autodetected
- Moved all actions to subfolders
- Split Remove Unit into 4 action files named after the actions.

Commit:
Merge branch 'master' into mission-api/dev

sorenmarkert committed:
Mission API: Add action specs (#8711)

### Work done

Added specs of all actions currently in the Mission API.

---------

Co-authored-by: Floris <ruwetuin@hotmail.com>
Co-authored-by: efrec <ecfrechette@gmail.com>

efrec committed:
Mission API: Remove unimplemented (#8785)

### Work done

Remove files and folders marked `yet_to_implement`.

Co-authored-by: efrec <quantumscrewdriver@gmail.com>

efrec committed:
Mission API: Fix loadout utilities bug (#8805)

### Work done

The codemod merge made changes on master but not on any other dev branches, including mission-api. Test missions currently will fail.

Co-authored-by: efrec <quantumscrewdriver@gmail.com>

Damgam committed:
Mission API - Move/Rotate Unit Actions (#8541)

- Added RotateUnit action 
Provide it with an angle or point on the map towards which the unit should be rotated
- Added MoveUnit action 
Provide it with point on the map towards which the unit should be teleported
Also allows rotating within the same action
- Added Direction Parameter Type which can accept both an angle parameter, as well as direction coordinates.

Commit:
Merge branch 'master' into mission-api/dev

WatchTheFort committed:
More action categories (#8872)

Split up the misc actions folder into more relevant subdirectories.

efrec committed:
Mission API: Build queue status (#8692)

### Work done

Patches the `ConstructionStarted` and `ConstructionFinished` triggers to ignore spawned and resurrected units that were never being built.

Adds unit tracking parameters to filter by the builder unit.

Converts placed build orders that become a build-assist so they are seen by `ConstructionStarted`. This only activates the trigger when the trigger filters on builderName or builderDefName; otherwise, the first to place activates, always.

Adds the `ContructionCanceled` and `ProductionCanceled` triggers to handle build orders that fail.

- The trigger `ConstructionCanceled` has use-case nuisances but I kept it for now: Any builder can complete another builder's task (some caveats), even an enemy, under the right circumstances.
- The trigger `ProductionCanceled` papers over the engine's split choice to cancel units in progress when the factory is captured but not when captured inside a factory. We consider the build to be canceled in both cases, plus when the factory is destroyed (and ofc when the unit itself is canceled).

Remaining `Production*` family triggers are produced elsewhere, PR tk.

### Testing

Specs and trigger scenario test.

Added specs to the existing triggers as well.

efrec committed:
Mission API: Construction progress (#8686)

### Work done

Adds the `ConstructionProgress` trigger, activated on reaching a buildProgress value on a candidate unit.

The candidates are checked against live state from a build-step summary callin. It is possible but would be unusual to reach a build%, destroy the unit with prejudice (freeing its unitID), and miss the completion event in this poll.

Alongside the trigger, this adds the `Fraction` parameter type to handle quantities from 0.0 to 1.0.

### Testing

Tests used are busted specs for construction_progress, validation, and
trigger load, and the runtime trigger tests.

Commit:
Merge branch 'master' into mission-api/dev

efrec committed:
Mission API: Unit spotted by sensor type (#8565)

### Work done

Added the `UnitDetected` and `UnitUndetected` triggers to extend the existing line-of-sight trigger to radar and seismic sensors. The previous `Spotted` triggers are retired.

Alongside these triggers are their requisite engine callins, two artificial callins (`UpdateContacts`, `DetectionUpdate`, need rename), shared trigger state (`SeismicContacts`, `DetectionLevels`), specs, and entries in the triggers example.

These callins are not added to the base game; we have no actual requirements for a "seismic detection state" outside mission scripting. So this code provides its own mark-and-sweep update path across all detection types and tackles trigger complexity, bringing it back down to a linear cost (sometimes cheaper). This can be improved once mission-api/dev and master converge but right now it would be dead code.

`UnitSeismicPing` needed a stateful approach to determining when a unit falls out of seismic detection. There are no engine events for "entering" or "leaving" seismic range, only a series of pings. My method uses a statistical process control with an anti-flapping latch. I'm not versed in signal processing, just SPC, so my version is just a statistical test.

efrec committed:
Mission API: merge two callins (#8902)

### Work done

Fix two GameFramePosts appearing. The github diff algo is just a drag on merges.

efrec committed:
Mission AI: Fix stale unit triggers test (#8903)

### Work done

Remove an old UnitSpotted test, superceded by UnitDetected with sensorTypes = { 'vision' }.

Commit:
add trigger unit_reclaimed, tests, specs

Ruwetuin committed:
Mission API - add action: unit targetable (#8548)

sorenmarkert committed:
Mission API: Unit detection fixes (#8910)

### Work done

- detection_levels no longer count the unit's own allyTeam when no sensorAllyTeam is named.
- Detection specs no longer rely on global state from other specs
- `construction_progress_spec.lua` no longer calls its previous/unused summary view

efrec committed:
Stylua pass on mission api + specs (#8908)

WatchTheFort committed:
Merge branch 'mission-api/dev' into mission-api/trigger/unit-reclaimed

efrec committed:
Mission API: Add trigger on production starting (#8904)

Adds the `ProductionStarted` trigger and its spec and tests.

Four factories "start building" by beginning an animation, at the moment. The Cortex bot lab has a significant delay, likely to be removed. The new trigger detects only the creation of the unit which is much more appropriate.

efrec committed:
Mission API: Unit received order (#8661)

### Work done

Adds three triggers for handling received commands:

- `UnitOrdered` on unit commands received and likely to execute by that unit. The command may be dropped within the engine, still. See notes in unit_ordered.
- `RallyOrdered` on factory commands received and unable to execute. These go in a special queue for built units (so must be `queueing = true`).
- `ProductionOrdered` on factory commands received to make units. These go in another queue, the factory queue.

Adds the `UnitCommand` callin and its dispatch to the Mission API triggers.

Adds the `Command` parameter type separately from `Orders`. It can accept an "any" and "build" qualifier, for now, to make working with mission scripts with no editor tolerable.

The trigger has a filter to ignore orders issued by actions. Call sites for `GiveOrder*` within the Mission API are fenced.

Commit:
Merge branch 'master' into mission-api/dev

efrec committed:
Merge branch 'mission-api/dev' into mission-api/trigger/unit-reclaimed

Commit:
missing }

Commit:
intercept reclaim_units actions

I'll draft a change on the engine side to accept weaponDefID and handle env weapon type behaviors, namely that reclaim=>no explosion, no wreck for this case, but probably needs to handle a lot of odd logic and may be rejected for baking that logic further into engine internals. Unsure.

tldr: Probably we should proceed with this. The engine course is the longer one.

Commit:
reset tests

Commit:
fromMission => ignoreMissionActions

efrec committed:
Mission API: Unit idled (#8724)

### Work done

Adds the `UnitIdled` and symmetric `UnitUnidled` triggers for detecting layabouts.

This circumvents a barely-usable engine callin, `:UnitIdle`, in favor of another summary callin, `UnitIdlePost`, which is a mark-and-sweep that we use to ignore sub-frame trigger flapping.

This does not yet mark factories or unit transfers. What idle contract belongs on just-built units is up in the air. The contradiction is apparent on factories: A just-built factory with no build queue should show idle. I'm unsure the reason to treat any other just-built unit as non-idle so hope we can fix this up soon.

Commit:
fix: emmylua complaint about no scriptPath

Commit:
stylua: api_missions

bandaid, meet your maker

efrec committed:
Mission API: Add unit reclaimed trigger (#8905)

### Work done

Adds the `UnitReclaimed` trigger and its tests and spec.

Uses the envDamageType to gate on reclaimed units, so has to check manually for the `ReclaimUnit` mission action, which registers as the plain `KilledByLua` type. There is an engine ticket to make this process easily customizable for Recoil games.

efrec committed:
stylua: Mission API triggers dispatch (#8947)

### Work done

Styling for api_missions_triggers plus minor reorganization because stylua does a bad job on a naive pass. We can make it look nice, we just have to try.
